File: Models/Recording.cs
using System;
using JetBrains.Annotations;

namespace Models
{
    /// <summary>
    /// Recording of a musical work.
    /// </summary>
    [PublicAPI]
    public record Recording
    {
        /// <summary>
        /// Id in the database.
        /// </summary>
        public int Id { get; init; }
        
        /// <summary>
        /// Filename of recording cover.
        /// It can be an album cover, or composite picture of several albums, or just some generic placeholder.
        /// </summary>
        /// <example>foo.webp</example>
        public string CoverName { get; init; } = "";
        
        /// <summary>
        /// Year when recording started. For single year either YearStart or YearFinish can be used.
        /// </summary>
        public int? YearStart { get; init; }
        
        /// <summary>
        /// Year when recording finished. For single year either YearStart or YearFinish can be used.
        /// </summary>
        public int? YearFinish { get; init; }

        /// <summary>
        /// List of performers associated with the recordings.
        /// </summary>
        public Performer[] Performers { get; init; } = Array.Empty<Performer>();
        
        /// <summary>
        /// Label which released the recording.
        /// </summary>
        /// <example>Ondine</example>
        public string? Label { get; init; }
        
        /// <summary>
        /// Length of the recording in minutes.
        /// </summary>
        public int Length { get; init; }

        /// <summary>
        /// List of streaming services which can stream the recording.
        /// </summary>
        public Streamer[] Streamers { get; init; } = Array.Empty<Streamer>();
    }
}

File: Site/Domain/Recording.fs
/// Business logic for Recordings.
module Site.Domain.Recording

open Site.Domain.Performer
open Site.Domain.Streamer
open Site.Postgres

/// Recording of a musical work.
type Recording =
    { id: int
      coverName: string
      yearStart: int option
      yearFinish: int option
      performers: Performer list
      label: string option
      length: int
      streamers: Streamer list }

/// Returns recordings of a given work.
let listRecordings (workId: int) : Async<string list> =
    let sql = "SELECT recordings_by_work(@WorkId) AS json"
    let parameters = [ "WorkId", Sql.int workId ] |> Some
    query(sql, parameters, jsonMapper)
    

